### PR TITLE
Extension api

### PR DESCRIPTION
**Pull request by kurtbrose**
WIP -- you can kind of see how the main glom loop just becomes a dispatcher

### CODE REVIEW

**Comment by kurtbrose** [comment]
(NOTE: still a work-in-progress)

basically, this is changing around the structure so that the main glom callback is a very simple dispatch loop:

```python
    def _glom(self, target, spec, context):
        # TODO: de-recursivize this
        # TODO: rearrange the branching below by frequency of use
        # recursive self._glom() calls should pass path=path to elide the current
        # step, otherwise add the current spec in some fashion
        context = context.new_child({})
        for type_, handler in self._spec_registry.items():
            if type_ is callable and callable(spec):
                return handler(spec, target, context)
            if type_ is not callable and isinstance(spec, type_):
                return handler(spec, target, context)
```

currently, the handlers dispatch table is an instance of the glommer; but, there's nothing stopping us from making it a part of context; if we did that, then you could have completely disjoint gloms pasted together into one big glom (by overriding the dispatch tables & etc in context for the sub-tree that wants to play by different rules)

here's what that dispatch table looks like currently:

```python
_DEFAULT_SPEC_REGISTRY = OrderedDict([
    (Inspect, Inspect._handler),
    (dict, _handle_dict),
    (list, _handle_list),
    (tuple, _handle_tuple),
    (_TType, _t_eval),  # NOTE: must come before callable b/c T is also callable
    (Call, Call._handler),
    (callable, lambda s, t, c: s(t)),
])
```

here's a very simple (but non-trivial) call-back for handling tuples:

```python
def _handle_tuple(spec, target, context):
    res = target
    for subspec in spec:
        res = context[HANDLE_CHILD](res, subspec, context)
        if not isinstance(subspec, list):
            context[Path] += [getattr(subspec, '__name__', subspec)]
    return res
```

here's the most complex one, the callback for `Inspect` -- it takes two new methods of the `Inspect` class to implement

(note, the Inspect code is possible to be separated out rather than scattered here and there)
```python
    def _handler(self, target, context):
        # stash the real handler under Inspect,
        # and replace the child handler with a trace callback
        context[Inspect] = context[HANDLE_CHILD]
        context[HANDLE_CHILD] = self._trace
        return context[HANDLE_CHILD](target, self.wrapped, context)

    def _trace(self, target, spec, context):
        if not self.recursive:
            context[HANDLE_CHILD] = context[Inspect]
        if self.echo:
            print('---')
            print('path:  ', context[Path] + [spec])
            print('target:', target)
        if self.breakpoint:
            self.breakpoint()
        try:
            ret = context[Inspect](target, self.wrapped, context)
        except Exception:
            if self.post_mortem:
                self.post_mortem()
            raise
        if self.echo:
            print('output:', ret)
            print('---')
        return ret
```

**Comment by kurtbrose** [comment]
the CI failures are doctests which I haven't figured out how to get running locally

**Comment by kurtbrose** [comment]
for the current registration / registry (sequence-walking and path-walking on targets) -- I'm thinking it would make sense to have a `SequenceHandler` and `PathHandler` -- or something akin to that

the job of these two classes would be to store the current registries and provide call-backs that can go into the spec-registry; sort of break down the current `_get_handler` into something that is owned on a spec-type by spec-type basis

**Comment by kurtbrose** [comment]
hmm... need to figure out how to get coverage.py to ignore chainmap_backport.py; I could just add `--omit chainmap_backport.py` to the coverage run in tox.ini I think; but I also see it referenced in another source file not sure which one codecov &etc works off of

maybe I need to modify both?

**Comment by mahmoud** [comment]
OK cool, looking at that last change, this is getting pretty close to shippable. I'm pretty sure there's a Gödelian issue/feature we can make by maintaining the invariant that `S[T] != S`, but instead `T` continues to reference its previous value. The motivation behind this is twofold:

* You already have `S`, so `S[T]` equaling `S` kind of doesn't add anything.
* You can continue iterating while referring to the scope, so: `[(S, T.method(S['some_context']))]` continues to work. This probably isn't a very common pattern, but for us, and some future subset of power users, this pattern will help avoid having to write extensions and reduce extraneous `Call` usage.

Does that make sense?

**Comment by kurtbrose** [comment]
(S, lambda scope: scope[T]) == original target (see unit test)

S[T] != S is very hard to interpret

e.g. spec is often a dict; target is often a dict
Trying to tell if the current Target is a "real" Target could be fraught

Can you express the behavior in terms of a concrete call to glom()?


On Jun 27, 2018 9:19 PM, "Mahmoud Hashemi" <notifications@github.com> wrote:

OK cool, looking at that last change, this is getting pretty close to
shippable. I'm pretty sure there's a Gödelian issue/feature we can make by
maintaining the invariant that S[T] != S, but instead T continues to
reference its previous value. The motivation behind this is twofold:

   - You already have S, so S[T] equaling S kind of doesn't add anything.
   - You can continue iterating while referring to the scope, so: [(S,
   T.method(S['some_context']))] continues to work. This probably isn't a
   very common pattern, but for us, and some future subset of power users,
   this pattern will help avoid having to write extensions and reduce
   extraneous Call usage.

Does that make sense?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/mahmoud/glom/pull/32#issuecomment-400906021>, or mute
the thread

File: glom/test/test_basic.py [hunk 1]
```diff
@@ -200,6 +200,14 @@ def test_spec_and_recursion():
 def test_scope():
     assert glom(None, S['foo'], scope={'foo': 'bar'}) == 'bar'
 
+    target = range(3)
+    spec = [(S, lambda S: S['multiplier'] * S[T])]
```
Comment (kurtbrose) [inline glom/test/test_basic.py]: I think usage of `S` as a local variable here is going to encourage confusion between the global spec `S`, and the local `scope` argument to `glom`

I'd like to propose the following convention for our docs &etc: `lambda scope:` where possible; `lambda s:` where writing out `scope` is too obtrusive.  Aliasing a global is usually considered bad practice, and then consider that it is just 10 characters to the left being used to mean something completely different I think this is particularly dangerous.

**Comment by kurtbrose** [comment]
oops -- so I added a line to `test_scope` `assert glom('hello', (S, lambda scope: scope[T])) == 'hello'`... but apparently somehow didn't get it pushed to github

this conversation makes a lot more sense now

(I wrote that test thinking it was going to fail and I was doing TDD... then it worked so I called it done)

**Comment by mahmoud** [comment]
Well, I think it's finally refactored, tested, and ready. Coverage is up and I'm merging, for a more extensible future :)

**Pull request merged**